arc: west: mdb: reverse the launch order for multi cores

IRISZZW opened the pull request:
The Inter-core Debug Unit (ICD) provides additional debug
assist features in multi-core scenarios.
In master core(core 0) initial stage, we will program ICD to halt
all other cores based on a halt occurring in one ore more core.
And all cores are in halt mode on reset, so we need to make
sure other slave cores have launched and in running mode
before we enable ICD in master core.

Currently we launch master first, Let's reverse the launch
order, launch master last, to make sure slave cores have
launched before we program and enable ICD.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

fix: #36467

IRISZZW commented:
without fix:
```
runners.mdb-hw: mdb -pset=1 -psetname=core0 '' -nooptions -nogoifmain -toggle=include_local_symbols=1 -digilent '' 'zephyr.elf'
runners.mdb-hw: mdb -pset=2 -psetname=core1 -prop=download=2 -nooptions -nogoifmain -toggle=include_local_symbols=1 -digilent '' 'zephyr.elf'
runners.mdb-hw: mdb -pset=3 -psetname=core2 -prop=download=2 -nooptions -nogoifmain -toggle=include_local_symbols=1 -digilent '' 'zephyr.elf'
runners.mdb-hw: mdb -pset=4 -psetname=core3 -prop=download=2 -nooptions -nogoifmain -toggle=include_local_symbols=1 -digilent '' 'zephyr.elf'
runners.mdb-hw: mdb -multifiles=core0,core1,core2,core3 -run '-cmd=-nowaitq run' -cmd=quit -cl
```
with fix:
```
runners.mdb-hw: mdb -pset=4 -psetname=core3 -prop=download=2 -nooptions -nogoifmain -toggle=include_local_symbols=1 -digilent '' zephyr.elf'
runners.mdb-hw: mdb -pset=3 -psetname=core2 -prop=download=2 -nooptions -nogoifmain -toggle=include_local_symbols=1 -digilent '' 'zephyr.elf'
runners.mdb-hw: mdb -pset=2 -psetname=core1 -prop=download=2 -nooptions -nogoifmain -toggle=include_local_symbols=1 -digilent '' 'zephyr.elf'
runners.mdb-hw: mdb -pset=1 -psetname=core0 '' -nooptions -nogoifmain -toggle=include_local_symbols=1 -digilent '' 'zephyr.elf'
runners.mdb-hw: mdb -multifiles=core3,core2,core1,core0 -run '-cmd=-nowaitq run' -cmd=quit -cl
```

ruuddw commented:
Relying on the start order by the debugger feels not very robust- can't the ICD init not be postponed till arch_start_cpu has been called for all CPUs? Though it looks like that would require adding a new arch callout in z_smp_init(), or some other mechanism to signal all cores have been brought up.

abrodkin commented:
> Relying on the start order by the debugger feels not very robust- can't the ICD init not be postponed till arch_start_cpu has been called for all CPUs? Though it looks like that would require adding a new arch callout in z_smp_init(), or some other mechanism to signal all cores have been brought up.

@ruuddw note this is only needed when we start Zephyr with JTAG preparing and kick-starting all cores in the cluster one-by-one. And since all operations conducted via JTAG are relatively slow, there's a huge difference in where a core may get to while others are not yet started.

In real-life all cores would be unhalted/started simultaneously and that problem won't happen.

And for the record, we've done the same fix for OpenOCD runner quite some time ago, see https://github.com/zephyrproject-rtos/zephyr/pull/22602/commits/57fea86925ed29faf405bd313f5afa66fec53943.

Now addressing your suggestion on fixing that problem in Zephyr's code... we did fix that in the Linux kernel with each slave core  re-programming ICD adding its own bit in the mask of "to be halted" cores, see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=f3205de98db2fc8083796dd5ad81b191e436fab8.
But in case of Linux we were trying to solve a bit another problem: running of the image capable of supporting up-to 4 cores on a cluster with say 1 or 2 cores or some of the cores halted.

In case of Zephyr this is most probably not a problem at all since we build an image for a particular configuration rather than universal something ;)

That said, to keep things simple let's fix a debugger commands invocation order rather than implementing more complex solutions in the target code.

abrodkin reviewed (changes requested):
A couple of notes:
1. The only change we need is this one:
    ```diff
    - mdb -multifiles=core0,core1,core2,core3 -run '-cmd=-nowaitq run' -cmd=quit -cl
    + mdb -multifiles=core3,core2,core1,core0 -run '-cmd=-nowaitq run' -cmd=quit -cl
    ```
    That's because all the previous invocations of `mdb` just create configurations for those `coreX` instances. But order of their execution is strictly set by how we mention them in `-multifiles`. That will this fix much simpler, compact and portable.

2. I'd add `ARConnect` to the name of ICD in the commit message to not confuse readers as otherwise there might be many others ICD (for example quite common is In-Circuit-Debugger) things in the world. Though you don't need to add `ARConnect` to all instances of ICD, but at least do it once, first time you refer to it.

ruuddw commented:
> note this is only needed when we start Zephyr with JTAG preparing and kick-starting all cores in the cluster one-by-one.

That is assuming all cores start/are taken out of reset at the same time in the SoC design. I think the purpose of the Zephyr 'arch_start_cpu' function is to allow systems to also work differently, where one CPU starts and manages the other CPUs, takes them out of reset or whatever mechanism to start.
If this synchronous start is something that all ARC HS SMP clusters share (i.e. other SoCs using HS SMP would also have a single reset starting all cores, individual cores don't have different 'halt-on-reset' values) then it is ok rely on this. At least, I think adding some warnings in the code initializing ICD on this assumption would make sense to avoid others debugging the same issue in the future.

abrodkin commented:
> If this synchronous start is something that all ARC HS SMP clusters share (i.e. other SoCs using HS SMP would also have a single reset starting all cores, individual cores don't have different 'halt-on-reset' values) then it is ok rely on this. At least, I think adding some warnings in the code initializing ICD on this assumption would make sense to avoid others debugging the same issue in the future.

It's not strictly an SoC feature, but rather a design decision for a particular product. I.e. in case of ARC each core might be configured to either run on reset or halt. So we may even in theory have cores 0 & 2 running on reset and cores 1 & 3 halted :)
What's more it might be a boot-ROM or whatever bootloader is executed before Zephyr who stops/starts certain cores. For example in case of HSDK only core 0 is configured to run on reset, while all the rest are halted. And when we boot Linux on HSDK in an autonomous way (true unmanned boot w/o debugger) it's U-Boot who sets an entry point & kick-starts all slave cores (again, slave cores start to execute Linux binary a bit ahead of the master, but that's just an implementation detail here).

That said, it's not an easy and straight-forward topic and a lot might go wrong. Still given relative simplicity of the improvement we did in the Linux kernel here (https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=f3205de98db2fc8083796dd5ad81b191e436fab8) we may in fact do the same here. But maybe as a complimentary improvement to this one. What do you think?

IRISZZW committed:
arc: west: mdb: reverse the launch order for multi cores

The ARConnect Inter-core Debug Unit (ICD) provides
additional debug assist features in multi-core scenarios.
In master core(core 0) initial stage, we will program ICD to halt
all other cores based on a halt occurring in one ore more core.
And all cores are in halt mode on reset, so we need to make
sure other slave cores have launched and in running mode
before we enable ICD in master core.

Currently we launch master first, Let's reverse the launch
order, launch master last, to make sure slave cores have
launched before we program and enable ICD.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

evgeniy-paltsev commented:
Looks like this need to be reverted as the proper fix is #36633
